File: benchmark/src/python/pidigits.py
# Pi digits — Machin: π = 16*arctan(1/5) - 4*arctan(1/239)
# Native Python int (unbounded). D=600. Golden: 2668_6766940513
import time

D = 600
GUARD = 5
SCALE = D + 10

def arctan_inv(x):
    res = 0
    term = 10**SCALE // x
    x2 = x * x
    sign = 1
    k = 0
    while True:
        divisor = 2 * k + 1
        t_k = term // divisor
        if t_k == 0:
            break
        if sign > 0:
            res += t_k
        else:
            res -= t_k
        sign = -sign
        term //= x2
        k += 1
    return res

start = time.time() * 1000

a5 = arctan_inv(5)
a239 = arctan_inv(239)
pi = 16 * a5 - 4 * a239

s = str(pi)
if len(s) < D + 10:
    s = s.zfill(D + 10)
s = s[:D]

ds = sum(int(c) for c in s)
lt = s[-10:]

end = time.time() * 1000
print(f"Result: {ds}_{lt}")
print(f"Time: {int(end - start)}ms")

File: benchmark/src/nodejs/pidigits.js
// Pi digits — Machin: π = 16*arctan(1/5) - 4*arctan(1/239)
// Node.js BigInt. D=600. Golden: 2668_6766940513

const D = 600;
const GUARD = 5;
const SCALE = D + 10;

function arctanInv(x) {
    let res = 0n;
    let term = (10n ** BigInt(SCALE)) / BigInt(x);
    const x2 = BigInt(x * x);
    let sign = 1;
    let k = 0;
    while (true) {
        const divisor = BigInt(2 * k + 1);
        const tk = term / divisor;
        if (tk === 0n) break;
        if (sign > 0) res += tk;
        else res -= tk;
        sign = -sign;
        term /= x2;
        k++;
    }
    return res;
}

const start = process.hrtime.bigint();

const a5 = arctanInv(5);
const a239 = arctanInv(239);
const pi = 16n * a5 - 4n * a239;

let s = pi.toString();
if (s.length < D + 10) s = s.padStart(D + 10, '0');
s = s.slice(0, D);

let ds = 0;
for (let i = 0; i < D; i++) ds += parseInt(s[i]);
const lt = s.slice(-10);

const end = process.hrtime.bigint();
const ms = Number((end - start) / 1000000n);
console.log(`Result: ${ds}_${lt}`);
console.log(`Time: ${ms}ms`);
